File: backend/demo/run_demo.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const BASE_DIR = path.join(__dirname, '..', '..', 'data', 'Fold1');

function ensureDir() {
  fs.mkdirSync(BASE_DIR, { recursive: true });
}

function writeSplit(filename, qidStart, nQueries = 5, docsPerQuery = 4) {
  const p = path.join(BASE_DIR, filename);
  const lines = [];
  for (let qi = qidStart; qi < qidStart + nQueries; qi++) {
    for (let di = 1; di <= docsPerQuery; di++) {
      const label = [0,1,2,3,4][Math.floor(Math.random()*5)];
      const feats = [];
      for (let fi = 1; fi <= 6; fi++) {
        const val = (Math.random()).toFixed(4);
        feats.push(`${fi}:${val}`);
      }
      const line = `${label} qid:${qi} ${feats.join(' ')} # docid=Q${qi}D${di}`;
      lines.push(line);
    }
  }
  fs.writeFileSync(p, lines.join('\n') + '\n', 'utf8');
}

async function postSearch(server = 'http://localhost:3000', query = 'wireless headphones') {
  const url = server.replace(/\/$/, '') + '/api/search';
  const body = { query, weights: {} };
  try {
    const res = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body)
    });
    const text = await res.text();
    console.log('Server response:', text);
  } catch (err) {
    console.error('Failed to call', url, err.message || err);
  }
}

async function main() {
  ensureDir();
  writeSplit('train.txt', 1, 6, 5);
  writeSplit('vali.txt', 1001, 3, 4);
  writeSplit('test.txt', 2001, 3, 4);
  console.log('Demo dataset generated in', BASE_DIR);

  console.log('\nAttempting example search request to dev server (http://localhost:3000)...');
  await postSearch();
}

main().catch(err => { console.error(err); process.exit(1); });

File: backend/demo/run_demo.py
#!/usr/bin/env python3
"""Run a quick demo:
- generate synthetic demo data
- optionally POST a sample search request to the running dev server (http://localhost:3000)
"""

import json
import time

from generate_demo_data import generate_all

DEFAULT_SERVER = "http://localhost:3000"


def post_search(server=DEFAULT_SERVER, query="wireless headphones"):
    import urllib.request

    url = server.rstrip("/") + "/api/search"
    data = json.dumps({"query": query, "weights": {}}).encode("utf-8")
    req = urllib.request.Request(url, data=data, headers={"Content-Type": "application/json"})
    try:
        with urllib.request.urlopen(req, timeout=5) as resp:
            body = resp.read().decode("utf-8")
            print("Server response:")
            print(body)
    except Exception as e:
        print(f"Failed to call {url}: {e}")


def main():
    print("Generating synthetic demo dataset...")
    generate_all()
    print("Waiting 0.5s for IO...")
    time.sleep(0.5)

    print("\nAttempting example search request to dev server (http://localhost:3000)...")
    post_search()


if __name__ == "__main__":
    main()
